include/geos/util/Interrupt.h

```diff
@@ -27,33 +27,55 @@ class GEOS_DLL Interrupt {
 public:
 
     typedef void (Callback)(void);
+    typedef void (ThreadCallback)(void*);
```

Comment:
I think we could simplify the design by keeping the old and new implementations separate: We could create a separate class to house the new context-specific interrupt implementation. This would likely also help with my earlier comment about how not all of the API surface that makes sense for the old API is relevant to the new API. My current thinking is that we could update the `GEOS_CHECK_FOR_INTERRUPTS()` definition to check for interrupts from both sources. Something along the lines of:

```
#define GEOS_CHECK_FOR_INTERRUPTS() geos::util::Interrupt::process(); geos::util::ThreadSpecificInterrupt::process();
```

Reply:
Additionally, I think this design would make it easier to one day drop the old interrupt API (something I hope we consider since my main interest in interruption stems from folks who use GEOSwift [reporting tooling-detected data races](https://github.com/GEOSwift/GEOSwift/issues/192) related to the old implementation)